typing/rec_check.ml

```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+            Guard
+      in
+      list expression exprs << array_mode
     | Texp_construct (_, desc, exprs) ->
       let access_constructor =
         match desc.cstr_tag with
-        | Cstr_extension (pth, _) -> Use.inspect (path env pth)
-        | _ -> Use.empty
+        | Cstr_extension (pth, _) ->
+          path pth << Dereference
+        | _ -> empty
       in
-      let use = match desc.cstr_tag with
-        | Cstr_unboxed -> (fun x -> x)
-        | Cstr_constant _ | Cstr_block _ | Cstr_extension _ -> Use.guard
+      let m' = match desc.cstr_tag with
+        | Cstr_unboxed ->
+          Return
+        | Cstr_constant _ | Cstr_block _ | Cstr_extension _ ->
+          Guard
       in
-      Use.join access_constructor (use (list expression env exprs))
+      join [
+        access_constructor;
+        list expression exprs << m'
+      ]
     | Texp_variant (_, eo) ->
-      Use.guard (option expression env eo)
+      (*
+        G |- e: m[Guard]
+        ------------------   -----------
+        G |- `A e: m         [] |- `A: m
+      *)
+      option expression eo << Guard
     | Texp_record { fields = es; extended_expression = eo;
                     representation = rep } ->
-        let use = match rep with
-          | Record_float -> Use.inspect
-          | Record_unboxed _ -> (fun x -> x)
+        let field_mode = match rep with
+          | Record_float -> Dereference
+          | Record_unboxed _ -> Return
           | Record_regular | Record_inlined _
-          | Record_extension -> Use.guard
+          | Record_extension -> Guard
         in
-        let field env = function
-            _, Kept _ -> Use.empty
-          | _, Overridden (_, e) -> expression env e
+        let field (_label, field_def) = match field_def with
+            Kept _ -> empty
+          | Overridden (_, e) -> expression e
         in
-        Use.join
-          (use (array field env es))
-          (option expression env eo)
+        join [
+          array field es << field_mode;
+          option expression eo << Dereference
+        ]
     | Texp_ifthenelse (cond, ifso, ifnot) ->
-        Use.(join (inspect (expression env cond))
-                (join
-                    (expression env ifso)
-                    (option expression env ifnot)))
+      (*
+        Gc |- c: m[Dereference]
+        G1 |- e1: m
+        G2 |- e2: m
+        ---
+        Gc + G1 + G2 |- if c then e1 else e2: m
+
+      Note: `if c then e1 else e2` is treated in the same way as
+      `match c with true -> e1 | false -> e2`
+      *)
+      join [
+        expression cond << Dereference;
+        expression ifso;
+        option expression ifnot;
+      ]
     | Texp_setfield (e1, _, _, e2) ->
-        Use.(join (inspect (expression env e1))
-              (inspect (expression env e2)))
+      (*
+        G1 |- e1: m[Dereference]
+        G2 |- e2: m[Dereference] (*TODO: why is this a dereference?*)
```

Comment:
Yes, I wanted @yallop to have a chance to comment on them, because he wrote the original implementation. I believe that using `Guard` here would be correct, but I wanted his opinion before making the check more liberal than the one we replace. (Other opinions are also welcome, of course.)

Reply:
It's a dereference because writing a value to a mutable field involves accessing the value.  Here's a program that fails with a segmentation fault if `Dereference` is weakened to `Return` here:

```ocaml
    type t = { mutable f: float }
    let g = { f = 0.0 }
    let rec x = (g.f <- y; ()) and y = 2.0
```

(You could add the program to the test suite.)